src/instructlab/model/train.py

```diff
@@ -521,20 +615,333 @@ def get_files(directory: str, pattern: str) -> list[str]:
+            backend=backends.VLLM,
+            enable_serving_output=False,
+        )
+        logger.debug("Judging mt-bench answers")
+        mt_bench_results: tuple = evaluator.judge_answers(model_serve_url)
+        ckpt_score: float = mt_bench_results[0]
+    finally:
+        if server is not None:
+            server.shutdown()
+
+    return ckpt_score
+
+
+def _evaluate_dir_of_checkpoints(
+    checkpoints_dir_path: pathlib.Path, eval_func: typing.Callable[..., float]
+) -> tuple[float, pathlib.Path]:
+    """Run eval_func on all model checkpoints in a directory, returning the best performing."""
+
+    # TODO: parallelize MMLU over available GPUs
+
+    results: list[typing.Tuple[float, pathlib.Path]] = []
+    for ckpt_path in checkpoints_dir_path.iterdir():
+        if ckpt_path.is_dir():
+            logger.debug(ckpt_path)
+            ckpt_score = eval_func(ckpt_path=ckpt_path)
             click.secho(
-                f"Error while executing training library {exc}",
+                f"CHECKPOINT EVALUATION: {ckpt_path} SCORED {ckpt_score}",
                 fg="red",
+                bg="cyan",
             )
-            raise click.exceptions.Exit(1)
+            results.append((ckpt_score, ckpt_path))
+
+    if not results:
+        raise RuntimeError(
+            "_evaluate_dir_of_checkpoints - No checkpoints were evaluated successfully at %s. No scores were recored."
+            % checkpoints_dir_path
+        )
+
+    ckpt_performance: list[tuple[float, pathlib.Path]] = sorted(results, reverse=True)
+
+    # (score, path)
+    return ckpt_performance[0]
+
+
+def _run_phased_training(
+    ctx,
+    train_args: TrainingArgs,
+    torch_args: TorchrunArgs,
+    base_dir: pathlib.Path,
+    phase1_data: pathlib.Path,
+    phase1_num_epochs: int | None,
+    phase1_samples_per_save: int | None,
+    phase2_data: pathlib.Path,
+    phase2_num_epochs: int | None,
+    phase2_samples_per_save: int | None,
+    mtbench_judge: pathlib.Path,
+) -> None:
+    p1_ckpt_dir: pathlib.Path = base_dir / "phase1" / "checkpoints"
+    p2_ckpt_dir: pathlib.Path = base_dir / "phase2" / "checkpoints"
+
+    if not p1_ckpt_dir.is_dir() and p2_ckpt_dir.is_dir():
+        raise FileNotFoundError(
+            "Paths to checkpoint storage for phase1 and phase2 must be directories."
+        )
+
+    if not phase1_data.exists() and phase2_data.exists():
+        raise FileNotFoundError(
+            "Data for both phase1 and phase2 must be specified for phased training."
+        )
+
+    if not mtbench_judge.resolve().exists():
+        raise FileNotFoundError(
+            f"MT-Bench model direcory could not be found: {mtbench_judge}\nMust be an absolute path to a model directory."
+        )
+
+    # -----------------------------------------------------------------------------
+    click.secho("Training Phase 1/2...", fg="cyan")
+
+    p1_train_args: TrainingArgs = train_args.model_copy(deep=True)
```

Comment:
same, I don't think you need the type here. `model_copy` returns `Self`; there are more unnecessary typing hints below, you may want to clean these up too.

Reply:
++